src/solver.rs

```diff
@@ -0,0 +1,134 @@
+//! Solvers for detecting known script kinds.
+
+use bounded_vec::BoundedVec;
+
+use crate::{
+    external::pubkey::PubKey,
+    op,
+    opcode::{push_value::LargeValue::PushdataBytelength, Evaluable, PushValue},
+    script, Opcode,
+};
+
+/// Parses a `script::PubKey` and detects standard scripts.
+///
+/// If successful, returns the script kind, and any relevant properties parsed from the
+/// script. For example, for a P2SH script, the result will contain the script hash; for
+/// P2PKH it will contain the key hash, etc.
+///
+/// Returns `None` if the script is non-standard.
+pub fn standard(script_code: &script::PubKey) -> Option<ScriptKind> {
+    match &script_code.0[..] {
+        // Pay-to-Script-Hash (P2SH)
+        [op::HASH160, Opcode::PushValue(PushValue::LargeValue(PushdataBytelength(v))), op::EQUAL] => {
+            v.as_slice()
+                .try_into()
+                .ok()
+                .map(|hash| ScriptKind::ScriptHash { hash })
+        }
+
+        // Pay-to-Public-Key-Hash (P2PKH)
+        [op::DUP, op::HASH160, Opcode::PushValue(PushValue::LargeValue(PushdataBytelength(v))), op::EQUALVERIFY, op::CHECKSIG] => {
+            v.as_slice()
+                .try_into()
+                .ok()
+                .map(|hash| ScriptKind::PubKeyHash { hash })
+        }
+
+        // Provably prunable, data-carrying output.
+        //
+        // So long as the script passes the `is_unspendable()` test, and all but the first
+        // byte passes the `is_push_only()` test, we don't care what exactly is in the
+        // script.
+        [op::RETURN, rest @ ..] => rest
+            .iter()
+            .map(|op| op.extract_push_value().ok().cloned())
+            // This `.collect()` is equivalent to `script::Opcode::is_push_only` for the
+            // subset of inputs consisting of entirely valid `PushValue`s, which is all we
+            // care about here.
+            .collect::<Option<_>>()
+            .map(|data| ScriptKind::NullData { data }),
+
+        // Legacy Pay-to-Public-Key (P2PK), which only occurred in early Zcash coinbase outputs.
+        [Opcode::PushValue(PushValue::LargeValue(PushdataBytelength(v))), op::CHECKSIG] => {
+            (v.len() == PubKey::SIZE || v.len() == PubKey::COMPRESSED_SIZE).then(|| {
+                ScriptKind::PubKey {
+                    data: v.clone().to_vec().try_into().expect("bounds checked"),
+                }
+            })
+        }
+
+        // Pay-to-Multi-Signature (P2MS)
+        [Opcode::PushValue(PushValue::SmallValue(required)), pubkeys @ .., Opcode::PushValue(PushValue::SmallValue(keys)), op::CHECKMULTISIG] =>
+        {
+            match (u8::try_from(required.to_num()), u8::try_from(keys.to_num())) {
+                (Ok(required @ 0..=16), Ok(keys @ 0..=16)) => {
```

Comment:
In standard.cpp, `IsSmallInteger` checks that the value is at least 1.
```suggestion
                (Ok(required @ 1..=16), Ok(keys @ 1..=16)) => {
```

Reply:
Good catch, thanks!